gapis/api/vulkan/frame_loop.go

```diff
@@ -404,14 +416,82 @@ func (f *frameLoop) buildStartEndStates(ctx context.Context, startState *api.Glo
 				f.descriptorSetLayoutDestroyed[descriptorSetLayout] = true
 			}
 
-			// TODO: Recreate destroyed resources.
-		}
+		// Command Buffers
+		case *VkAllocateCommandBuffers:
+			vkCmd := cmd.(*VkAllocateCommandBuffers)
+			cmdBuffers := vkCmd.PCommandBuffers().MustRead(ctx, vkCmd, currentState, nil)
+			f.commandBufferAllocated[cmdBuffers] = true
+			log.D(ctx, "Command buffer %v allcoated.", cmdBuffers)
+		case *VkFreeCommandBuffers:
+			vkCmd := cmd.(*VkFreeCommandBuffers)
+			cmdBufCount := vkCmd.CommandBufferCount()
+			if cmdBufCount == 0 {
+				break
+			}
+			cmdBufs := vkCmd.PCommandBuffers().Slice(0, uint64(cmdBufCount), currentState.MemoryLayout).MustRead(ctx, cmd, currentState, nil)
+			for _, cmdBuf := range cmdBufs {
+				// The command buffer deleted in this call was created during loop, no action needed.
+				log.D(ctx, "Command buffer %v freed", cmdBufs)
+				if _, ok := f.commandBufferAllocated[cmdBuf]; ok {
+					delete(f.commandBufferAllocated, cmdBuf)
+				} else { // The command buffer deleted in this call was not created during loop, need to back up it
+					f.commandBufferFreed[cmdBuf] = true
+				}
+			}
+		case *VkBeginCommandBuffer:
+			vkCmd := cmd.(*VkBeginCommandBuffer)
+			cmdBuf := vkCmd.CommandBuffer()
+			f.commandBufferRecorded[cmdBuf] = true
+			log.D(ctx, "Command buffer %v beginned", cmdBuf)
+
+			// If this command buffer is allocated during loop then it is reset by the re-allocation step.
+			if _, ok := f.commandBufferAllocated[cmdBuf]; ok {
+				break
+			}
+			// If this command buffer has not been submitted before no action needed.
+			if _, ok := f.commandBufferNotRecorded[cmdBuf]; !ok {
+				break
+			}
+			// Only backup the fist time it changes
+			if _, ok := f.commandBufferChanged[cmdBuf]; !ok {
+				f.commandBufferChanged[cmdBuf] = true
+			}
 
-		if err := cmd.Mutate(ctx, cmdId, currentState, nil, f.watcher); err != nil {
-			return fmt.Errorf("%v: %v: %v", cmdId, cmd, err)
+		case *VkResetCommandBuffer:
+			vkCmd := cmd.(*VkResetCommandBuffer)
+			cmdBuf := vkCmd.CommandBuffer()
+			log.D(ctx, "Command buffer %v resetted", cmdBuf)
+			// If this command buffer is allocated during loop then it is reset by the re-allocation step.
+			if _, ok := f.commandBufferAllocated[cmdBuf]; ok {
+				break
+			}
+			// If this command buffer has not been submitted before no action needed.
+			if _, ok := f.commandBufferNotRecorded[cmdBuf]; !ok {
+				break
+			}
+			// Only backup the fist time it changes
+			if _, ok := f.commandBufferChanged[cmdBuf]; !ok {
+				f.commandBufferChanged[cmdBuf] = true
+			}
+		case *VkQueueSubmit:
+			vkCmd := cmd.(*VkQueueSubmit)
+			vkCmd.Extras().Observations().ApplyReads(currentState.Memory.ApplicationPool())
```

Comment:
I think this is done outside the loop in the current code? So no need to reapply the observations.